### PR TITLE
feat(portal): Add REST API

### PR DESCRIPTION
**Pull request by bmanifold**
Why:

* In order to manage a large number of Firezone Sites, Resources, Policies, etc... a REST API is needed as clicking through the UI is too time consuming, as well as prone to error.  By providing a REST API Firezone customers will be able to manage things within their Firezone accounts with code.

### CODE REVIEW

File: elixir/apps/api/lib/api/plugs/auth.ex [hunk 1]
```diff
@@ -11,9 +11,16 @@ defmodule API.Plugs.Auth do
...
+        # |> send_resp(401, Jason.encode!(%{"error" => "invalid_access_token"}))
+        # |> halt()
+
+        # TODO: BRIAN - Confirm that this change won't break anything with the clients or gateways
         conn
-        |> put_resp_content_type("application/json")
-        |> send_resp(401, Jason.encode!(%{"error" => "invalid_access_token"}))
+        |> put_status(401)
+        |> Phoenix.Controller.put_view(json: API.ErrorJSON)
+        |> Phoenix.Controller.render(:"401")
```
Comment (bmanifold) [inline elixir/apps/api/lib/api/plugs/auth.ex:23]: I made this change to allow the REST API errors to be uniform, but I left this note here for myself because I realized after making it that I wasn't sure if the client and gateways relied on the specific error message.

**Comment by bmanifold** [comment]
~I still need to finish up Actor Group Memberships, as well as potentially the create/update for Identities, but I figured since I was going to be out the next two days I'd push up what I have and get some feedback.~
Done

File: elixir/apps/api/lib/api/controllers/fallback_controller.ex [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    |> put_status(:not_found)
+    |> put_view(json: API.ErrorJSON)
+    |> render(:"404")
+  end
+
+  def call(conn, {:error, :unauthorized}) do
+    conn
+    |> put_status(403)
+    |> put_view(json: API.ErrorJSON)
+    |> render(:"403")
```
Comment (jamilbk) [inline elixir/apps/api/lib/api/controllers/fallback_controller.ex]: Just curious if you meant `401` here?

```suggestion
    |> put_status(401)
    |> put_view(json: API.ErrorJSON)
    |> render(:"401")
```

File: elixir/apps/api/lib/api/controllers/gateway_controller.ex [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  # Show a specific Gateway
+  def show(conn, %{"id" => id}) do
+    with {:ok, gateway} <- Gateways.fetch_gateway_by_id(id, conn.assigns.subject) do
+      render(conn, :show, gateway: gateway)
+    end
+  end
+
+  # Create a new Gateway Token
```
Comment (jamilbk) [inline elixir/apps/api/lib/api/controllers/gateway_controller.ex]: I don't feel super strongly about this, but what are your thoughts on moving the token lifecycle endpoints to `gateway_groups`? That might align better with how users are able to use them with our current architecture.

E.g. `/gateway_groups/tokens` for the CRUD actions.

Comment (jamilbk) [reply elixir/apps/api/lib/api/controllers/gateway_controller.ex]: Actually I'm not sure if we save tokens in the DB today?

File: elixir/apps/api/lib/api/controllers/gateway_json.ex [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  end
+
+  defp data(%Gateways.Gateway{} = gateway) do
+    %{
+      id: gateway.id,
+      name: gateway.name,
+      ipv4: gateway.ipv4,
+      ipv6: gateway.ipv6
+    }
+  end
```
Comment (jamilbk) [inline elixir/apps/api/lib/api/controllers/gateway_json.ex]: I know this is going to be tricky, and if so we can save for a later point, but just in case it's not -- I think it would be super cool to have online/offline status in here too

File: elixir/apps/api/lib/api/plugs/auth.ex (hunk 1 above)
Comment (jamilbk) [reply elixir/apps/api/lib/api/plugs/auth.ex:23]: We should be ok here. I think we just check the code and not the specific message.

See https://github.com/firezone/firezone/blob/main/rust/phoenix-channel/src/lib.rs#L88

@conectado can you confirm?

File: elixir/apps/domain/lib/domain/actors/group.ex [hunk 5]
```diff
@@ -20,8 +20,9 @@ defmodule Domain.Actors.Group do
     # ref https://github.com/firezone/firezone/issues/2162
     has_many :actors, through: [:memberships, :actor]
 
-    field :created_by, Ecto.Enum, values: ~w[system identity provider]a
+    field :created_by, Ecto.Enum, values: ~w[actor identity provider system]a
```
Comment (jamilbk) [inline elixir/apps/domain/lib/domain/actors/group.ex:23]: 👍

**Review by jamilbk — COMMENTED** [review]
Looking good. I know this will probably start a debate, but any thoughts on using a header to specify the API version instead of a path param?


Here are some previous discussions on the topic:

https://github.com/firezone/firezone/issues/2041#issuecomment-1715983219
https://github.com/firezone/firezone/issues/2203
https://github.com/firezone/firezone/issues/2209

File: elixir/apps/api/lib/api/controllers/fallback_controller.ex (hunk 2 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/fallback_controller.ex]: Yep, I always mix this up with HTTP statuses.  I always forget that `Unauthorized` in HTTP is really `Unauthenticated` and `Forbidden` is really `Unauthorized`.

File: elixir/apps/api/lib/api/controllers/gateway_controller.ex (hunk 3 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/gateway_controller.ex]: I don't feel super strong about this either.  Just to make sure we're on the same page, you're suggesting the following endpoints?:
```
POST    /v1/gateway_groups/:gateway_group_id/tokens    <- To create tokens
DELETE  /v1/gateway_groups/:gateway_group_id/tokens/:token_id    <- To delete/revoke a gateway group token
```
And then these would be the only endpoints nested under `gateway_groups`:
```
GET     /v1/gateway_groups/:gateway_group_id/gateways 
GET     /v1/gateway_groups/:gateway_group_id/gateways/:id
DELETE  /v1/gateway_groups/:gateway_group_id/gateways/:id
  ```

>Actually I'm not sure if we save tokens in the DB today?

We do save tokens in the DB ([migration to add tokens table](https://github.com/firezone/firezone/blob/main/elixir/apps/domain/priv/repo/migrations/20230920172332_add_tokens.exs))

File: elixir/apps/api/lib/api/controllers/gateway_json.ex (hunk 4 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/gateway_json.ex]: Yeah, I initially put it in, but then realized I was going to have to deal with the Phoenix Presence stuff and figured I'd push that off until I had some time to chat with @AndrewDryga

File: elixir/apps/api/lib/api/controllers/gateway_controller.ex (hunk 3 above)
Comment (jamilbk) [reply elixir/apps/api/lib/api/controllers/gateway_controller.ex]: Yeah I think so

Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/gateway_controller.ex]: 👍 Cool, that's what I ended up implementing.  I also added in:
```
DELETE /v1/gateway_groups/:gateway_group_id/tokens
```
That will delete/revoke all tokens for a given Gateway Group

File: elixir/apps/api/lib/api/controllers/actor_controller.ex [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    with {:ok, actor} <- Actors.fetch_actor_by_id(id, conn.assigns.subject) do
+      render(conn, :show, actor: actor)
+    end
+  end
+
+  # Create a new Actor
+  def create(conn, %{"actor" => params}) do
+    subject = conn.assigns.subject
+
+    with {:ok, account} <- Domain.Accounts.fetch_account_by_id(subject.account.id, subject),
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/controllers/actor_controller.ex]: We already have `subject.account`, so we don't need to load it 2nd time

File: elixir/apps/api/lib/api/controllers/actor_controller.ex [hunk 7]
```diff
@@ -0,0 +1,56 @@
+defmodule API.ActorController do
+  alias Domain.Actors
+  import API.ControllerHelpers
+  use API, :controller
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/controllers/actor_controller.ex:2]: The order should be `use`, `import`, `alias`

File: elixir/apps/api/lib/api/controllers/controller_helpers.ex [hunk 8]
```diff
@@ -0,0 +1,23 @@
+defmodule API.ControllerHelpers do
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/controllers/controller_helpers.ex]:
```suggestion
defmodule API.Pagination do
```

and then it would be better to jus use `Pagination.params_to_list_opts(params)` everywhere instead of doing `import API.ControllerHelpers`. 

`import` adds compile-time dependency and pollutes the namespace of the module, plus it's not super clear where the function is coming from where the module has a bunch of `import` or `use`'s.

File: elixir/apps/api/lib/api/controllers/fallback_controller.ex (hunk 2 above)
Comment (AndrewDryga) [reply elixir/apps/api/lib/api/controllers/fallback_controller.ex]: @bmanifold same here 🤣

File: elixir/apps/api/lib/api/controllers/actor_group_json.ex [hunk 9]
```diff
@@ -0,0 +1,35 @@
+defmodule API.ActorGroupJSON do
+  alias Domain.Actors
+  alias Domain.Repo.Paginator.Metadata
+
+  @doc """
+  Renders a list of Actor Groups.
+  """
+  def index(%{actor_groups: actor_groups, metadata: metadata}) do
+    %{data: for(actor_group <- actor_groups, do: data(actor_group))}
+    |> Map.put(:metadata, metadata(metadata))
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/controllers/actor_group_json.ex]: This might be a bit easier to read:
```suggestion
    %{
      data: Enum.map(actor_groups, &data/1), 
      metadata: metadata(metadata)
    }
```

File: elixir/apps/api/lib/api/controllers/actor_group_json.ex [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  defp data(%Actors.Group{} = actor_group) do
+    %{
+      id: actor_group.id,
+      name: actor_group.name
+    }
+  end
+
+  defp metadata(%Metadata{} = metadata) do
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/controllers/actor_group_json.ex]: I think this can be moved to `API.Paginator` so you won't have aliases in all JSON modules that have `index/1` function.

File: elixir/apps/api/lib/api/controllers/gateway_json.ex [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  Render a single Gateway
+  """
+  def show(%{gateway: gateway}) do
+    %{data: data(gateway)}
+  end
+
+  @doc """
+  Render a Gateway Token
+  """
+  def token(%{gateway_token: token}) do
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/controllers/gateway_json.ex]: This also should be in `gateway_group_json.ex`.

File: elixir/apps/api/lib/api/controllers/gateway_json.ex (hunk 4 above)
Comment (AndrewDryga) [reply elixir/apps/api/lib/api/controllers/gateway_json.ex]: We can do that here too, don't have anything against it. All you need to do it ass `preload: online?` to the fetch/list functions and include it here.

File: elixir/apps/api/lib/api/controllers/actor_controller.ex (hunk 6 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/actor_controller.ex]: Ah, yeah, that makes sense.  Will get this updated

File: elixir/apps/api/lib/api/controllers/actor_controller.ex (hunk 7 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/actor_controller.ex:2]: I always forget this order.  Thank you for the reminder.

File: elixir/apps/api/lib/api/controllers/resource_controller.ex [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  end
+
+  def show(conn, %{"id" => id}) do
+    with {:ok, resource} <- Resources.fetch_resource_by_id(id, conn.assigns.subject) do
+      render(conn, :show, resource: resource)
+    end
+  end
+
+  def create(conn, %{"resource" => params}) do
+    with attrs <- set_param_defaults(params),
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/controllers/resource_controller.ex]: You don't want to put `=` as the first element of with, that will generate an `case` clause like this:
```
case set_param_defaults(params) do
  attrs -> ..rest of with block..
end
```
```suggestion
    attrs = set_param_defaults(params)
    
    with 
```

File: elixir/apps/api/test/api/controllers/actor_controller_test.exs [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+    account = Fixtures.Accounts.create_account()
+    actor = Fixtures.Actors.create_actor(type: :api_client, account: account)
+
+    %{
+      account: account,
+      actor: actor
+    }
+  end
+
+  describe "index" do
```
Comment (AndrewDryga) [inline elixir/apps/api/test/api/controllers/actor_controller_test.exs]: We should add `/arity` in the of describe blocks, it makes it easier to find tests by grep'ing and it just a general convention for Elixir:
```suggestion
  describe "index/1" do
```

File: elixir/apps/api/test/api/controllers/actor_group_controller_test.exs [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+      assert json_response(conn, 200) == %{
+               "data" => %{
+                 "id" => actor_group.id,
+                 "name" => actor_group.name
+               }
+             }
+    end
+  end
+
+  describe "create" do
```
Comment (AndrewDryga) [inline elixir/apps/api/test/api/controllers/actor_group_controller_test.exs]: It would be also nice to add tests for validation errors (at least what happens when `params` are empty).

File: elixir/apps/api/test/api/controllers/example_controller_test.exs [hunk 15]
```diff
@@ -4,7 +4,7 @@ defmodule API.ExampleControllerTest do
   describe "echo/2" do
     test "returns error when not authorized", %{conn: conn} do
       conn = post(conn, "/v1/echo", %{"message" => "Hello, world!"})
-      assert json_response(conn, 401) == %{"error" => "invalid_access_token"}
+      assert json_response(conn, 401) == %{"errors" => %{"detail" => "Unauthorized"}}
```
Comment (AndrewDryga) [inline elixir/apps/api/test/api/controllers/example_controller_test.exs]: Would you expect `errors` to return a list or a map? From naming I would assume a list will be rendered, not a single item as a map.
```suggestion
      assert json_response(conn, 401) == %{"errors" => %{"detail" => "Unauthorized"}}
```

File: elixir/apps/api/test/api/controllers/gateway_controller_test.exs [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+                 "name" => gateway.name,
+                 "ipv4" => Domain.Types.IP.to_string(gateway.ipv4),
+                 "ipv6" => Domain.Types.IP.to_string(gateway.ipv6)
+               }
+             }
+
+      assert {:error, :not_found} ==
+               Gateway.Query.not_deleted()
+               |> Gateway.Query.by_id(gateway.id)
+               |> Repo.fetch(Gateway.Query)
```
Comment (AndrewDryga) [inline elixir/apps/api/test/api/controllers/gateway_controller_test.exs]: This is just a style preference but I would do something like this:
```suggestion
      assert gateway = Repo.get(Gateway, gateway.id)
      assert gateway.deleted_at
```
WDYT?

File: elixir/apps/api/test/api/controllers/gateway_group_controller_test.exs [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+
+      assert count == 3
+      assert limit == 50
+      assert is_nil(next_page)
+      assert is_nil(prev_page)
+
+      data_ids = Enum.map(data, & &1["id"]) |> MapSet.new()
+      gateway_group_ids = Enum.map(gateway_groups, & &1.id) |> MapSet.new()
+
+      assert MapSet.equal?(data_ids, gateway_group_ids)
```
Comment (AndrewDryga) [inline elixir/apps/api/test/api/controllers/gateway_group_controller_test.exs]: What if we add a helper to `ConnCase` called `equal_ids?(list, list)` which will do all of this? This code is duplicated in a lot of places.

File: elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    |> execute("")
+  end
+
+  def change do
+    for table_name <- @table_names do
+      alter table(table_name) do
+        add(:created_by_actor_id, :uuid)
+      end
+
+      migrate_data(table_name)
```
Comment (AndrewDryga) [inline elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs:34]: I would also add a CHECK constraint that will require certain fields depending on `created_by` enum field.

File: elixir/apps/domain/lib/domain/policies/policy/changeset.ex [hunk 19]
```diff
@@ -14,8 +14,8 @@ defmodule Domain.Policies.Policy.Changeset do
     |> cast_embed(:conditions, with: &Domain.Policies.Condition.Changeset.changeset/3)
     |> changeset()
     |> put_change(:account_id, subject.account.id)
-    |> put_change(:created_by, :identity)
-    |> put_change(:created_by_identity_id, subject.identity.id)
+    |> put_change(:created_by_actor_id, subject.actor.id)
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/policies/policy/changeset.ex]: Let's move this to `put_created_by` too.

File: elixir/apps/domain/lib/domain/clients/authorizer.ex [hunk 20]
```diff
@@ -14,6 +14,13 @@ defmodule Domain.Clients.Authorizer do
     ]
   end
 
+  def list_permissions_for_role(:api_client) do
+    [
+      manage_own_clients_permission(),
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/clients/authorizer.ex:19]: Not related to this PR: I hate the naming I did here 😢

File: elixir/apps/api/test/api/controllers/example_controller_test.exs (hunk 15 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/example_controller_test.exs]: I just went off of what the Phoenix generator created and figured that's what the norm was for the community.  By default here's what gets generated for the errors template:
```
defmodule MyWebApp.ErrorJSON do
  # If you want to customize a particular status code,
  # you may add your own clauses, such as:
  #
  # def render("500.json", _assigns) do
  #   %{errors: %{detail: "Internal Server Error"}}
  # end

  # By default, Phoenix returns the status message from
  # the template name. For example, "404.json" becomes
  # "Not Found".
  def render(template, _assigns) do
    %{errors: %{detail: Phoenix.Controller.status_message_from_template(template)}}
  end
end
```

I'm open to changing things, but just unsure of what type of effort would be involved if we moved away from the default.

File: elixir/apps/api/test/api/controllers/actor_group_controller_test.exs (hunk 14 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/actor_group_controller_test.exs]: Yeah, good call.  Will add those types of tests in.

File: elixir/apps/api/test/api/controllers/actor_controller_test.exs (hunk 13 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/actor_controller_test.exs]: Sounds good 👍

File: elixir/apps/api/lib/api/controllers/gateway_json.ex (hunk 11 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/gateway_json.ex]: 👍 Yeah, I've already moved it in my local branch after @jamilbk suggested moving the token creation under `gateway_groups` rather than `gateways`.  I'll double check that I've removed it from this file though.

File: elixir/apps/api/test/api/controllers/gateway_controller_test.exs (hunk 16 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/gateway_controller_test.exs]: Yeah, that's nice.  Will update the delete tests to match that.

File: elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs (hunk 18 above)
Comment (bmanifold) [reply elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs:34]: Gotcha, that makes sense.  I'm guessing that enum value would be `created_by` = `identity` in this case?

File: elixir/apps/domain/lib/domain/policies/policy/changeset.ex (hunk 19 above)
Comment (bmanifold) [reply elixir/apps/domain/lib/domain/policies/policy/changeset.ex]: 👍 Sounds good

File: elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs (hunk 18 above)
Comment (AndrewDryga) [reply elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs:34]: Yeah, and `actor` if identity_id is null.

File: elixir/apps/api/test/api/controllers/example_controller_test.exs (hunk 15 above)
Comment (AndrewDryga) [reply elixir/apps/api/test/api/controllers/example_controller_test.exs]: There should be no efforts, we can style out API responses the way we like

File: elixir/apps/api/lib/api/controllers/actor_controller.ex (hunk 7 above)
Comment (AndrewDryga) [reply elixir/apps/api/lib/api/controllers/actor_controller.ex:2]: To remember you can think "what is more polluting for the module":
- `use` can add functions via macro, aliases, imports and pretty much anything else, so it comes first;
- `import` adds functions to module namespace and a compile-time dependency, so comes second;
- `alias` just adds a module alias, third;
- `require` adds a compile-time dependency so that macros can be used but doesn't pollute the namespace at all, comes last.

File: elixir/apps/api/test/api/controllers/example_controller_test.exs (hunk 15 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/example_controller_test.exs]: Sorry, I should have been a little more clear on my last sentence.  Is there anything we miss out on by going outside the Phoenix generated template?

Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/example_controller_test.exs]: Also, is it going to get strange with Ecto change set errors?  Here's an example of what currently gets returned if there are multiple errors in a changeset:
```
{
  "errors": {
    "provider_identifier": [
      "can't be blank"
    ],
    "provider_identifier_confirmation": [
      "email does not match"
    ]
  }
}
```

If we're going to make `errors` a list, would we need to have a Keyword List in some instances and a List in others?

File: elixir/apps/api/test/api/controllers/actor_group_controller_test.exs (hunk 14 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/actor_group_controller_test.exs]: Ah, wait, don't I already have those tests.  Line 113 in this file for example.

File: elixir/apps/api/lib/api/plugs/auth.ex (hunk 1 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/plugs/auth.ex:23]: @conectado, just wanted to bump this confirmation

File: elixir/apps/api/test/api/controllers/example_controller_test.exs (hunk 15 above)
Comment (AndrewDryga) [reply elixir/apps/api/test/api/controllers/example_controller_test.exs]: @bmanifold good catch with changesets. It's pretty hard to change things like this one later so maybe we can just do a minimal structuring, maybe something like this:

```
# one-off error, like unathorized
%{error: %{reason: Phoenix.Controller.status_message_from_template(template)}}
# example
%{error: %{reason: "unathorized"}}

# changesets
%{error: %{reason: "bad_request", validation_errors: %{...}}}
```
or something along those lines?

This is not the most important thing though and we can also leave it as is if naming and restructuring everything will take too much time right now.

File: elixir/apps/api/test/api/controllers/actor_group_controller_test.exs (hunk 14 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/actor_group_controller_test.exs]: While I did have tests for empty attrs under the given attribute key, I went back and added tests with no params/body at all.

File: elixir/apps/api/test/api/controllers/example_controller_test.exs (hunk 15 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/example_controller_test.exs]: Error responses should now be in the form recommended

File: elixir/apps/api/test/api/controllers/gateway_group_controller_test.exs (hunk 17 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/gateway_group_controller_test.exs]: Done

File: elixir/apps/api/lib/api/controllers/resource_controller.ex (hunk 12 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/resource_controller.ex]: I made this change and I can see why in this case it wouldn't be necessary to have that in the `with` clause, but I'm slightly confused about the `with` clause generating a `case` clause.  I'm guessing I just need to read up on `with` a bit more.  Is there a specific set of docs talking about this that I could read?

Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/resource_controller.ex]: I also wanted to clarify something for my own edification. You mentioned not having a `=` in the first element of a `with` statement.  The original code I had used a `<-`, is that the same as using a `=` and if so, how is that different from something like:
```
with {:ok, foo} <- do_something() do
  ...block of code here...
end
```

File: elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs (hunk 18 above)
Comment (bmanifold) [reply elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs:34]: As I tried to work this CHECK out I realized that all of the Enum values aren't the same for every table.  Would you prefer a simple check that makes sure the `created_by` field isn't `identity` if the `created_by_identity_id` is NULL or would you prefer an individual customized CHECK per table?

File: elixir/apps/api/lib/api/controllers/actor_group_membership_controller.ex [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+      render(conn, :index, actors: actors, metadata: metadata)
+    end
+  end
+
+  # Update Actor Group Memberships
+  def update(
+        conn,
+        %{"actor_group_id" => actor_group_id, "memberships" => params}
+      ) do
+    add = Map.get(params, "add", [])
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/controllers/actor_group_membership_controller.ex:104]: Would it be better if we just accept a list of `%{actor_id: ...}` maps like changeset expects it? I think this will be more helpful with terraform provider that won't need to compare results and generate diffs and instead just apply to the state it expects. This is also a bit easier for syncs because third-party system won't need to compute the diff, but will be a bit harder when third-party system wants to do one-off update.

File: elixir/apps/api/lib/api/controllers/identity_controller.ex [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    params =
+      Map.put_new(
+        params,
+        "provider_identifier_confirmation",
+        Map.get(params, "provider_identifier")
+      )
+
+    with {:ok, actor} <- Domain.Actors.fetch_actor_by_id(actor_id, subject),
+         {:ok, provider} <- Auth.fetch_provider_by_id(provider_id, subject),
+         true = valid_provider?(provider),
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/controllers/identity_controller.ex]: This will crash the code and API consumer won't have a proper error message:
```
iex(1)> with true = false do
...(1)>   :ok
...(1)> end
** (MatchError) no match of right hand side value: false
    (stdlib 6.0) erl_eval.erl:652: :erl_eval.expr/6
    iex:1: (file)
```

Just FYI this will compile to:
```
true = false
:ok
```

File: elixir/apps/api/test/api/controllers/example_controller_test.exs [hunk 23]
```diff
@@ -4,7 +4,7 @@ defmodule API.ExampleControllerTest do
   describe "echo/2" do
     test "returns error when not authorized", %{conn: conn} do
       conn = post(conn, "/v1/echo", %{"message" => "Hello, world!"})
-      assert json_response(conn, 401) == %{"error" => "invalid_access_token"}
+      assert json_response(conn, 401) == %{"error" => %{"reason" => "Unauthorized"}}
```
Comment (AndrewDryga) [inline elixir/apps/api/test/api/controllers/example_controller_test.exs]: This controller should be deleted, I left it there just as example that you can use when building the API

File: elixir/apps/api/test/api/controllers/gateway_controller_test.exs [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+    gateway_group = Fixtures.Gateways.create_group(%{account: account})
+
+    %{
+      account: account,
+      actor: actor,
+      gateway_group: gateway_group
+    }
+  end
+
+  describe "index/2" do
```
Comment (AndrewDryga) [inline elixir/apps/api/test/api/controllers/gateway_controller_test.exs:17]: There are no tests for `create/2`.

File: elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs (hunk 18 above)
Comment (AndrewDryga) [reply elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs:34]: I think we can do one like this for all the tables:
```
(created_by = `identity` AND created_by_actor_id IS NOT NULL AND created_by_identity_id IS NOT NULL)
OR (created_by = `actor` AND created_by_actor_id IS NOT NULL AND created_by_identity_id IS NULL)
OR (created_by = `provider` AND provider_id IS NOT NULL AND created_by_actor_id IS NULL AND created_by_identity_id IS NULL)
OR (created_by = `system` AND provider_id IS NULL AND created_by_actor_id IS NULL AND created_by_identity_id IS NULL)
```

File: elixir/apps/api/test/api/controllers/gateway_controller_test.exs (hunk 24 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/gateway_controller_test.exs:17]: Yeah, there's no `create` function for the `gateway_controller`, because you shouldn't be able to create a gateway through the API right?

File: elixir/apps/api/test/api/controllers/example_controller_test.exs (hunk 23 above)
Comment (bmanifold) [reply elixir/apps/api/test/api/controllers/example_controller_test.exs]: 👍 Wasn't sure whether you wanted that left in or not, but I'll get it removed.

File: elixir/apps/api/lib/api/controllers/actor_group_membership_controller.ex (hunk 21 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/actor_group_membership_controller.ex:104]: Yeah, I see what you're saying.  I was thinking more in the one-off mindset when writing this.

What if we allowed both the current way and the way you're suggesting?

File: elixir/apps/api/lib/api/controllers/identity_controller.ex (hunk 22 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/identity_controller.ex]: Ah, yeah, that makes sense now that you point it out.  How would you handle this?  Would you just have an explicit check inside the `with` block that verifies that the provider is a `manual` provisioner or is there a slicker/more elegant way of handling this?

File: elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs (hunk 18 above)
Comment (bmanifold) [reply elixir/apps/domain/priv/repo/migrations/20240618210715_add_created_by_actor.exs:34]: Unfortunately, I don't think that will work for a couple of reasons.  The biggest being that we can't reference columns that don't exist on the table.  The second reason (at least in the local Dev DB, not sure if prod and staging are different) is that which ever value is in `created_by` doesn't guarantee that all of the other `created_by_<something>_id` fields will be NULL.

The second reason I need to dig in to a little bit more to see if that's just a side effect of how we seed our Dev DBs or if it actually ends up in that state in staging/prod.

File: elixir/apps/api/lib/api/controllers/gateway_json.ex (hunk 4 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/gateway_json.ex]: 👍 Added the `preload: :online?`

File: elixir/apps/api/test/api/controllers/gateway_controller_test.exs (hunk 24 above)
Comment (AndrewDryga) [reply elixir/apps/api/test/api/controllers/gateway_controller_test.exs:17]: @bmanifold ah, I was expecting there we would have a way to create a token, but now realized it's for the group. Sorry.

File: elixir/apps/api/lib/api/controllers/identity_controller.ex (hunk 22 above)
Comment (AndrewDryga) [reply elixir/apps/api/lib/api/controllers/identity_controller.ex]: You can either check for `:manual` value right there or just do `true <- valid_provider?(provider)` and add an `else` clause that will convert `false` to something meaningful for the client.

File: elixir/apps/api/lib/api/controllers/actor_group_membership_controller.ex (hunk 21 above)
Comment (AndrewDryga) [reply elixir/apps/api/lib/api/controllers/actor_group_membership_controller.ex:104]: That sounds good, maybe we can use different HTTP verb/path for that so that you can't mix both in the same request? Eg. send whole list in `PUT` and add/remove for `PATCH`?

Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/actor_group_membership_controller.ex:104]: Yeah, separating them by HTTP verb would match up great.

File: elixir/apps/api/lib/api/controllers/resource_controller.ex (hunk 12 above)
Comment (AndrewDryga) [reply elixir/apps/api/lib/api/controllers/resource_controller.ex]: @bmanifold the `with` clause just compiles down to a bunch of `case`s:
```
with :foo <- foo(),
      :bar <- bar() do
  :ok
else
  :fiz -> {:error, :fiz11}
  :fix -> {:error, :fix}
end
```
is the same as:
```
case foo() do
   :foo -> 
   case bar() do
       :bar -> :ok
      :fiz -> {:error, :fiz11}
      :fix -> {:error, :fix}
    end
  :fiz -> {:error, :fiz11}
  :fix -> {:error, :fix}
end
```

So there is no point to write this:
```
with foo <- foo() do
...code
end
= because it goes down to this =>
case foo do 
   foo -> ...code
end
= which can be just => 
foo = foo()
...code
```

**Review by AndrewDryga — APPROVED** [review]

File: elixir/apps/api/lib/api/controllers/resource_controller.ex (hunk 12 above)
Comment (bmanifold) [reply elixir/apps/api/lib/api/controllers/resource_controller.ex]: This makes sense.  I really appreciate the detailed explanation 🙏
